File: ims-main/backend/app/utils/ledger_display.py
"""库存流水/汇总页面展示分组（与前端 ledgerDisplay.js 一致）。"""
from app.models.enums import RETURN_CONDITION_ORDER, StockCondition

OTHER_INBOUND_TYPE_CODES = [c.value for c in RETURN_CONDITION_ORDER] + [
    StockCondition.RETURNED_FROM_BORROW.value,
]
OTHER_OUTBOUND_TYPE_CODES = ["SCRAPPED", "RND", "SAMPLE", "TRIAL", "REPAIR", "DEPT_PROCUREMENT", "BORROWED"]

LEDGER_INBOUND_DISPLAY_HEADERS = ["入库-正常", "入库-其他入库"]
LEDGER_OUTBOUND_DISPLAY_HEADERS = [
    "出库-售出-线上",
    "出库-售出-线下",
    "出库-准售出",
    "出库-赠送",
    "出库-其他出库",
]


def _sum_type_qty(type_qty: dict[str, int] | None, codes: list[str]) -> int:
    type_qty = type_qty or {}
    return sum(int(type_qty.get(code, 0)) for code in codes)


def inbound_display_values(inbound_by_type: dict[str, int] | None) -> list[int]:
    """入库展示列：正常、其他入库。"""
    inbound_by_type = inbound_by_type or {}
    normal = int(inbound_by_type.get(StockCondition.NEW.value, 0))
    other = _sum_type_qty(inbound_by_type, OTHER_INBOUND_TYPE_CODES)
    return [normal, other]


def outbound_display_values(outbound_by_type: dict[str, int] | None) -> list[int]:
    """出库展示列：售出-线上、售出-线下、准售出、赠送、其他出库。"""
    outbound_by_type = outbound_by_type or {}
    other = _sum_type_qty(outbound_by_type, OTHER_OUTBOUND_TYPE_CODES)
    return [
        int(outbound_by_type.get("SOLD", 0)),
        int(outbound_by_type.get("SOLD_OFFLINE", 0)),
        int(outbound_by_type.get("PRESOLD", 0)),
        int(outbound_by_type.get("GIFTED", 0)),
        other,
    ]

File: ims-main/frontend/src/utils/ledgerDisplay.js
import {
  OUTBOUND_TYPE_MAP,
  OUTBOUND_TYPE_ORDER,
  RETURN_CONDITION_ORDER,
  STOCK_CONDITION_MAP,
} from '@/constants/enums'

/** 其他入库：各类退回（含借用归还） */
export const OTHER_INBOUND_TYPE_CODES = [...RETURN_CONDITION_ORDER, 'RETURNED_FROM_BORROW']

/** 其他出库：除售出/准售出/赠送外的出库类型 */
export const OTHER_OUTBOUND_TYPE_CODES = ['SCRAPPED', 'RND', 'SAMPLE', 'TRIAL', 'REPAIR', 'DEPT_PROCUREMENT', 'BORROWED']

/** 流水明细：入库类型行（对应 inventory_sku_daily_ledger 入库列） */
export const LEDGER_INBOUND_DETAIL_ITEMS = [
  { code: 'NEW', label: STOCK_CONDITION_MAP.NEW },
  ...OTHER_INBOUND_TYPE_CODES.map((code) => ({
    code,
    label: STOCK_CONDITION_MAP[code] || code,
  })),
]

/** 流水明细：出库类型行（对应 inventory_sku_daily_ledger 出库列） */
export const LEDGER_OUTBOUND_DETAIL_ITEMS = [
  ...OUTBOUND_TYPE_ORDER.map((code) => ({
    code,
    label: OUTBOUND_TYPE_MAP[code] || code,
  })),
  { code: 'BORROWED', label: OUTBOUND_TYPE_MAP.BORROWED },
]

export const OTHER_INBOUND_TOOLTIP = `其他入库包含：${OTHER_INBOUND_TYPE_CODES.map((code) => STOCK_CONDITION_MAP[code] || code).join('、')}`

export const OTHER_OUTBOUND_TOOLTIP = `其他出库包含：${OTHER_OUTBOUND_TYPE_CODES.map((code) => OUTBOUND_TYPE_MAP[code] || code).join('、')}`

export function buildLedgerInboundDetailRows(typeMap) {
  return LEDGER_INBOUND_DETAIL_ITEMS.map(({ code, label }) => ({
    label,
    qty: typeMap?.[code] || 0,
  }))
}

export function buildLedgerOutboundDetailRows(typeMap) {
  return LEDGER_OUTBOUND_DETAIL_ITEMS.map(({ code, label }) => ({
    label,
    qty: typeMap?.[code] || 0,
  }))
}

export function sumTypeQty(typeMap, codes) {
  return codes.reduce((sum, code) => sum + (typeMap?.[code] || 0), 0)
}

export function ledgerInboundNormalQty(typeMap) {
  return typeMap?.NEW || 0
}

export function ledgerInboundOtherQty(typeMap) {
  return sumTypeQty(typeMap, OTHER_INBOUND_TYPE_CODES)
}

export function ledgerOutboundSoldQty(typeMap) {
  return typeMap?.SOLD || 0
}

export function ledgerOutboundSoldOfflineQty(typeMap) {
  return typeMap?.SOLD_OFFLINE || 0
}

export function ledgerOutboundPresoldQty(typeMap) {
  return typeMap?.PRESOLD || 0
}

export function ledgerOutboundGiftedQty(typeMap) {
  return typeMap?.GIFTED || 0
}

export function ledgerOutboundOtherQty(typeMap) {
  return sumTypeQty(typeMap, OTHER_OUTBOUND_TYPE_CODES)
}

/** 其他入库 hover 分项（仅数量 > 0） */
export function otherInboundBreakdownLines(typeMap) {
  return OTHER_INBOUND_TYPE_CODES.map((code) => ({
    label: STOCK_CONDITION_MAP[code] || code,
    qty: typeMap?.[code] || 0,
  })).filter((item) => item.qty > 0)
}

/** 其他出库 hover 分项（仅数量 > 0） */
export function otherOutboundBreakdownLines(typeMap) {
  return OTHER_OUTBOUND_TYPE_CODES.map((code) => ({
    label: OUTBOUND_TYPE_MAP[code] || code,
    qty: typeMap?.[code] || 0,
  })).filter((item) => item.qty > 0)
}
